contracts/dual_token/DualStakingToken.sol

```diff
@@ -0,0 +1,427 @@
+
+    // Service registry address
+    address public immutable serviceRegistry;
+    // Staking token address (except for OLAS)
+    address public immutable stakingToken;
+    // Service staking instance address
+    address public immutable stakingInstance;
+
+    // Required staking token amount
+    uint256 public stakingTokenAmount;
+    // Staking token ratio to OLAS rewards in 1e18 form
+    uint256 public rewardRatio;
+    // Staking token contract balance
+    uint256 public balance;
+    // Staking token available rewards
+    uint256 public availableRewards;
+    // Owner address
+    address public owner;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Mapping of staker account address => Staker info struct
+    mapping(uint256 => StakerInfo) public mapStakerInfos;
+    /// Mapping of staked OLAS service multisigs
+    mapping(address => bool) public mapMutisigs;
+
+    /// @dev DualStakingToken constructor.
+    /// @param _serviceRegistry Service registry address.
+    /// @param _stakingToken Staking token address.
+    /// @param _stakingInstance Service staking instance address.
+    /// @param _rewardRatio Staking token ratio to OLAS rewards in 1e18 form.
+    constructor(
+        address _serviceRegistry,
+        address _stakingToken,
+        address _stakingInstance,
+        uint256 _rewardRatio
+    ) {
+        // Check for zero addresses
+        if (_serviceRegistry == address(0) || _stakingToken == address(0) || _stakingInstance == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Check for zero values
+        if (_rewardRatio == 0) {
+            revert ZeroValue();
+        }
+
+        serviceRegistry = _serviceRegistry;
+        stakingToken = _stakingToken;
+        stakingInstance = _stakingInstance;
+
+        rewardRatio = _rewardRatio;
+
+        // Calculate staking token amount based on staking instance service information
+        uint256 numAgentInstances = IStaking(_stakingInstance).numAgentInstances();
+        uint256 minStakingDeposit = IStaking(_stakingInstance).minStakingDeposit();
+        // Total service deposit = minStakingDeposit + minStakingDeposit * numAgentInstances
+        stakingTokenAmount = (minStakingDeposit * (1 + numAgentInstances) * _rewardRatio) / 1e18;
+
+        owner = msg.sender;
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @notice The balance is always greater or equal the amount, as follows from the Base contract logic.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal {
+        // Update the contract balance
+        balance -= amount;
+
+        SafeTransferLib.safeTransfer(stakingToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Changes contract owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external {
```

Comment:
Should not have an owner. Must always be governance free as for Olas staking too